internal/oauthtoken/file.go

```diff
@@ -0,0 +1,86 @@
+// Copyright 2024 EngFlow Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package oauthtoken
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/fs"
+	"os"
+	"path/filepath"
+
+	"golang.org/x/oauth2"
+)
+
+// FileStore persists tokens in JSON files in a directory, one file per
+// token. The tokens are stored unecrypted. This implementation may be used
+// when Keyring cannot be used, for example, in a CI environment when there is
+// no encrypted keyring.
+type FileStore struct {
+	dir string
+}
+
+var _ LoadStorer = (*FileStore)(nil)
+
+func NewFileTokenStore(dir string) (*FileStore, error) {
+	if err := os.MkdirAll(dir, 0700); err != nil {
+		return nil, fmt.Errorf("creating token directory: %w", err)
+	}
+	return &FileStore{dir: dir}, nil
+}
+
+func (f *FileStore) Load(cluster string) (_ *oauth2.Token, err error) {
+	defer func() {
+		if err != nil {
+			err = fmt.Errorf("loading token for cluster %s: %w", cluster, err)
```

Comment:
Love the defer trick! I might steal this one :)

nit: I prefer `%q` when recording string params in errors; I've seen the root-cause bug being an empty string enough times, which isn't as obvious when the param is recorded with `%s`

Reply:
I'm a big fan of this trick. Please steal and use it everywhere! Around Go 1.13, there was a serious proposal to add an error handling construct to the language because error handling was one of the larger pain points in user surveys. But ultimately, the proposal was rejected because the construct seemed somewhat redundant with `defer`. The pattern here was considered a good alternative.

Reply:
Also, changed `%s` to `%q`.